pkgs/top-level/all-packages.nix

```diff
@@ -12306,7 +12306,8 @@ with pkgs;
 
   signal-cli = callPackage ../applications/networking/instant-messengers/signal-cli { };
 
-  inherit (callPackage ../applications/networking/instant-messengers/signal-desktop {}) signal-desktop signal-desktop-beta;
+  inherit (callPackage ../applications/networking/instant-messengers/signal-desktop {})
+    signal-desktop signal-desktop-beta signal-desktop-apple-emoji signal-desktop-beta-apple-emoji;
```

Comment:
Can't you do
```
signal-desktop-apple-emoji = signal-desktop.override { apple-emoji = true; };
```
?

Reply:
If I use

```nix
  inherit (callPackage ../applications/networking/instant-messengers/signal-desktop {}) signal-desktop signal-desktop-beta;

  signal-desktop-apple-emoji = signal-desktop.override { apple-emoji = true; };

  signal-desktop-beta-apple-emoji = signal-desktop-beta.override { apple-emoji = true; };
```

I get the error

```nix
error:
       … while calling a functor (an attribute set with a '__functor' attribute)

         at /nix/store/l0idbmripjzp8374jwyap7l85viwci4r-source/pkgs/top-level/all-packages.nix:12311:32:

        12310|
        12311|   signal-desktop-apple-emoji = signal-desktop.override { apple-emoji = true; };
             |                                ^
        12312|

       … while calling a functor (an attribute set with a '__functor' attribute)

         at /nix/store/l0idbmripjzp8374jwyap7l85viwci4r-source/lib/customisation.nix:144:43:

          143|       # Re-call the function but with different arguments
          144|       overrideArgs = mirrorArgs (newArgs: makeOverridable f (overrideWith newArgs));
             |                                           ^
          145|       # Change the result of the function call by applying g to it

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: function 'anonymous lambda' called with unexpected argument 'apple-emoji'

       at /nix/store/l0idbmripjzp8374jwyap7l85viwci4r-source/pkgs/applications/networking/instant-messengers/signal-desktop/signal-desktop.nix:1:1:

            1| { callPackage }:
             | ^
            2| callPackage ./generic.nix { } rec {
```

I think it's because there's two layers of `callPackage` indirection: first, `signal-desktop/default.nix` runs `callPackage ./signal-desktop.nix { };`, and then  `signal-desktop.nix` runs `callPackage ./generic.nix { }`. The inner function in `generic.nix` is the one that needs to be overridden, but `.override` seems to see only the outer function (whose only argument is `callPackage`). Not sure if there's a way around this.

https://github.com/NixOS/nixpkgs/blob/fbca5e745367ae7632731639de5c21f29c8744ed/pkgs/applications/networking/instant-messengers/signal-desktop/default.nix#L3-L7

https://github.com/NixOS/nixpkgs/blob/fbca5e745367ae7632731639de5c21f29c8744ed/pkgs/applications/networking/instant-messengers/signal-desktop/signal-desktop.nix#L1-L8

https://github.com/NixOS/nixpkgs/blob/fbca5e745367ae7632731639de5c21f29c8744ed/pkgs/applications/networking/instant-messengers/signal-desktop/generic.nix#L55-L60

I can't just add the `apple-emoji` flag to `signal-desktop.nix` because I can't get `nix-update` to update two hashes in the same file, since `nix-update` only takes an attribute, and not an arbitrary nix expression (if I could ask it to update `signal-desktop.override { apple-emoji = true; }`, then that might work).

Reply:
That would result in the copyright‐violating `.deb` being cached on `cache.nixos.org` for the cleaned package which is not okay; unfortunately it has to be cleaned in the same derivation as the download due to the fact that every distribution Hydra builds gets cached.

Sorry for not reviewing this yet, was busy all day. I will try to take a closer look tomorrow.